### ISSUE TITLE
As a user, I would like PRs to be "waiting on author" after a certain number of reviews

### ISSUE THREAD

**Issue by abinoda** [pullreminders/backlog#103]
Requested by @justinko:

> We require 2 approvals per pull request. Even after a pull has 2 approvals, I still get notified (because I'm requested and have not approved). It would be nice if there was a setting to specify the "less than number of approvals" to trigger a notification.

Requested by @jamesdo:

> Any shot there's a way to only do team alerts in the event the pull request doesn't have a minimum required amount of reviews? For our team, posting about an approved PR in the team notification is unnecessary as it's not truly actionable. We're good to merge upon any approval. The real friction point is when no one has reviewed a specific PR. Those are the things we like visibility into.

Requested by @endrec:

> We always add multiple reviewers to pull requests, and a single approval is enough. For example, we have a PR right now, where there are 3 reviewers, A, B and C. A and B both requested changes. The Team reminder shows this PR as waiting for C, though in reality, it is waiting for it's author to implement those changes. There is already a setting to "Remind authors after reviews", but it only works when all reviewer has finished their reviews. Would it be possible to add a "Number of reviews needed" option there, so if there are at least that much, it would go to the author?

Requested by @tinder-hemanthprasad:

> We require 2 reviews for PRs. Do you think we can "Remind authors after reviews" after 2 reviews?

**Comment by abinoda** [comment]
@jamesdo @endrec I've been thinking about the idea of adding a `Number of reviews needed` option for `Reminder authors after reviews`. My concern is around what happens when there are multiple rounds of review. For example, if the `Number of reviews needed` were set to `1`,

1. Author opens a pull request and assigns Reviewer A, Reviewer B, and Reviewer C
2. Reviewer B requests changes
3. Pull request is now "Waiting on author" since it has one review
4. Author makes changes, then requests another review from Reviewer B
5. Pull request is still "Waiting on author" because it has a review

I'm not sure how to solve for this, and considering if the two alternatives below are better solutions:

1. One option is instead of assigning individual reviewers, [assign reviews to a GitHub Team](https://blog.github.com/2017-06-28-requesting-reviews-from-repository-teams/). Pull Reminders handles notifications for team review requests, and when someone leaves a review, the review request for the team is cancelled and the pull request goes to "Waiting on author" until another review is requested.
2. If using GitHub Teams isn't a good option, another option would be to create a [GitHub Action](https://github.com/actions/) that automatically cancels remaining pending review requests whenever someone leaves a review (the exact logic of how many reviews or approvals triggers this could be configured). This would create a very clean workflow where after each round of review, the author could explicitly re-request reviews they'd like. I would be happy to create an open-source GitHub Action template for this that you could use and customize.

**Comment by endrec** [comment]
Creating a GitHub Action (which I never heard of before :blush: ) is an interesting idea, the biggest problem I can see that it would need to be added to every repo individually, I would prefer something centrally configured.

Multiple rounds would not be a problem in our case, as we tend to dismiss previous reviews when changes were made.

**Comment by jamesdo** [comment]
More than number of reviews, I believe number of approvals would be more useful. Regardless of whether or not a PR is waiting for review from a tagged person, it can be ignored if, for example, 2 team members have given explicit approval.

Reducing the time to _mergability_, which for our team is typically defined by # of explicit approvals.

And I _think_ me and @endrec may have a slightly different problem set here. Our team isn't so concerned with the state of requested changes.

**Comment by abinoda** [comment]
@jamesdo @endrec Thanks both of you for your feedback. It sounds like a solution here would be to add a `Number of approvals` settings as @jamesdo suggested, instead of a `Number of reviews` setting. If the setting is only applicable to approvals, then the problem I described earlier regarding multiple rounds of reviews is easily solved by [dismissing approvals](https://help.github.com/articles/dismissing-a-pull-request-review/) manually or automatically (e.g., GitHub has settings to automatically dismiss reviews after new commits).

How does this sound to you guys?

**Comment by endrec** [comment]
Yes, I agree with @jamesdo, we have a slightly different set of problems. Adding a `Number of Approvals` setting solves one of these problems.

In my opinion, notifying the author instead of the team, when changes are requested, also reduces the *time to mergability*, as the author can start adding the requested changes sooner.

**Comment by abinoda** [comment]
> In my opinion, notifying the author instead of the team, when changes are requested, also reduces the time to mergability, as the author can start adding the requested changes sooner.

Makes sense. So let me get this straight... 

@jamesdo is asking for "point reminders at the author when there's one approval, regardless of whether or not there are separate requested changes". 

And @endrec, you are asking for "point reminders when there's one review, regardless of whether it is an approval, requested change, or comments".

> Multiple rounds would not be a problem in our case, as we tend to dismiss previous reviews when changes were made.

@endrec  Can you 100% confirm that all review types are dismissed (approvals, requested changes, comments) when adding new commits or manually dismissing? Otherwise it could be tricky for Pull Reminders to follow multiple rounds of review.

**Comment by endrec** [comment]
I'd use a bit different wording to summarise @jamesdo's and our use case:

Remind the author (*not* the team) when
- there is at least the set number of approvals (different teams might require different number of approvals) **or**
- there is one or more requested change(s)

I can confirm that new commits should dismiss existing reviews of any kind. (We are working towards the 100%.)

**Comment by abinoda** [comment]
@endrec @jamesdo Gotcha, thanks for the help here. I'm going to move forward with this soon after I get a couple things in progress now out the door.

**Comment by abinoda** [comment]
Just had this same thing requested by @justinko:

> We require 2 approvals per pull request. Even after a pull has 2 approvals, I still get notified (because I'm requested and have not approved). It would be nice if there was a setting to specify the "less than number of approvals" to trigger a notification.

**Comment by abinoda** [comment]
@justinko Just shipped the first iteration of this – look under "More filters" for "My DM settings" and there's a way to exclude PRs with a specified number of non-dismissed approvals. Let me know how it works!

**Comment by abinoda** [comment]
@jamesdo @endrec @tinder-hemanthprasad I've been working on adding the `Number of reviews needed` setting for team reminders. For now I'm moving forward with the solution shown in **Code sample 1** below which I think should work for 95% of cases, but I wanted to get your feedback since this is a difficult problem.

As discussed earlier, this feature is difficult because Pull Reminders needs to be able to handle cases where there are multiple rounds of review. This means Pull Reminders needs to be able to determine which reviews should be counted toward the `Number of reviews needed` versus which should be considered stale. @endrec had suggested leveraging [GitHub's dismissed reviews](https://blog.github.com/2016-10-12-dismissing-reviews-on-pull-requests/) feature to determine the number of non-stale reviews. However, there are two issues I've identified with this:

1. GitHub treats comments as reviews (they nullify review requests). And these reviews can't be dismissed.
2. @endrec's team may be an exception but I think it is rare for teams to dismiss reviews as part of their workflow. I don't want to build this feature in a way that requires teams to manually dismiss reviews.

Based on this, I think that the best way to determine if a review is stale or not is by checking whether or not there is currently a review request for the same reviewer. For example, if Kelly reviews my PR then I create another review request for Kelly, then we would consider Kelly's review stale.

Here's a pseudo-code version of this logic:

```ruby
# Code sample 1

def needs_review?(pr)
  reviews = pr.non_dismissed_reviews.where('reviewer_id NOT IN (?)', pr.requested_reviewers)
  
  return pr.has_review_requests? && reviews.count < settings.needed_reviews
end
```

The main shortcoming I've identified with this implementation is that it breaks down if different reviewers are assigned across review rounds. For example, if Kelly and Jimmy review my PR then I request reviews from Jimmy and Tom, Kelly's review is not considered stale (unless it is dismissed) since I didn't request another review from Kelly.

### Possible alternative logic

An alternative approach I thought of was to simply treat reviews as stale based on new review requests. So if Kelly and Jimmy review my PR then I request new reviews from Kelly, Jimmy, and Tom, Kelly and Jimmy's reviews are considered stale.

Here's what this logic looks like:

```ruby
# Code sample 2 

def needs_review?(pr)
  reviews = pr.non_dismissed_reviews.where('submitted > ?', pr.last_review_request.created)
  
  return pr.has_review_requests? && reviews.count < settings.needed_reviews
end
```

The problem with this logic is when that it treats all past reviews as stale even if they shouldn't be. For example, if Kelly approves my PR and Jimmy requests changes, and then I request a review from Jimmy, Kelly's approval is considered stale since it happened before the latest review request.

One workaround to this could be to make an exception for approvals so they aren't considered stale unless the exact same reviewer is requested again. This would be baking in an assumption, though:

```ruby
# Code sample 3

def needs_review?(pr)
  approvals = pr.non_dismissed_approvals.where('reviewer_id NOT IN (?)', pr.requested_reviewers)  
  other_reviews = pr.non_dismissed_reviews.where('submitted > ?', pr.last_review_request.created)
  
  return pr.has_review_requests? && (approvals + other_reviews).count < settings.needed_reviews
end
```

### Other notes

On a completely separate note, @endrec suggested that there could be two separate options fo `Number of reviews needed` and `Number of approvals needed`. To keep things simple I'm hoping `Number of reviews needed` is enough.

**Comment by tinder-hemanthprasad** [comment]
- We sometimes have a lot of discussion on the PRs. If each comment is considered a review it makes it complicated. We generally don't add request reviews to encourage everyone in the team to review. 
- We look for 2 approvals to get the PR merged. One `ok` (anyone) and `ok-approved` (Senior developer or Module owners). We use labels to identify what's needed for the PR.

**Comment by abinoda** [comment]
This is live 🚀 – we've added a "Needed Reviews" setting to control when a PR should be considered "reviewed" and be assigned back to the author for action. This is helpful if your team typically assigns multiple reviewers to PRs but only require one or two reviews to consider PRs "reviewed".

<img width="682" alt="screen shot 2018-12-26 at 1 11 15 pm" src="https://user-images.githubusercontent.com/50083/50454847-c40f3080-090f-11e9-8e87-01f11c42bd86.png">

**Comment by abinoda** [comment]
Marking this as complete, but @endrec @jamesdo would appreciate feedback on this!

**Comment by endrec** [comment]
@abinoda Sorry for not responding earlier, but the last week was extremely busy for us. So far it seems the implemented solution covers our needs, thanks for your great job. :)

**Comment by abinoda** [comment]
@endrec Thanks for reporting back!

**Comment by ccpost** [comment]
@abinoda This is great! I love the distinction between whether action is on the reviewer(s) or the author.

I'm having a bit of trouble with getting action back on the reviewers on a particular PR, though. The flow went (I have it set to "After one review" since we just need one approval):
- PR opened -> A + B requested as reviewers (they are both also code owners on some files)
- A + B both left comments with no approval or changes requested
- Author updated PR with more commits
- A (me) noticed that PullReminders still thought action was back on the author. A removed A + B from review requests and added A + B back as reviewers.
- PullReminders still seems to think action is back on the author.

The last few things in the PR timeline look like this:
```
@A A removed their request for review 9 minutes ago
 @A A self-requested a review 8 minutes ago
 @A A removed request for B and A 6 minutes ago
 @A A requested review from A and B 6 minutes ago
```

This is a private repository, so if you want to look further on your end I can send you an email.

**Comment by abinoda** [comment]
Hi @ccpost – so Pull Reminders should work with the series of events you described. If reviewers A and B are requested, then both of them respond with comments, their review requests are fulfilled so the PR should flip to "Waiting on Author". Then, if reviewers A and B are requested again, the PR should go back to "Waiting on Reviewer".

Want to see if you can try the workflow I just described and reproduce the issue on your end, and then email me?

If it's helpful at all, here's the function in Pull Reminders that determines if a PR needs review from a reviewer or is waiting on an author:
```ruby
if @rule.include_reviewed_prs? && @rule.needed_reviews
  pr.has_review_requests? && pr.non_stale_review_count < @rule.needed_reviews
else
  pr.has_review_requests?
end
```

And here's the SQL fragment used to determine `non_stale_review_count`:

```sql
SELECT COUNT(DISTINCT(qr.github_user_id))
FROM github_reviews qr
JOIN github_users qru on qr.github_user_id = qru.id
WHERE qr.github_pr_id = github_prs.id
AND qr.dismissed IS FALSE
AND qru.github_id NOT IN (
  SELECT user_github_id
  FROM requested_reviewers qrr
  WHERE qrr.github_pr_id = github_prs.id
)
```

As you can see from the SQL, it looks for non-dismissed reviews but checks them against the current requested reviewers and doesn't count reviews by reviewers who are currently assigned.

**Comment by NejcZupec** [comment]
@abinoda We also love the distinction whether an action is on the author or the reviewers. However, we're facing a similar issue as @ccpost described.

We also require just one approval from all reviewers and use the following workflow setting:
<img width="413" alt="screenshot 2019-01-30 at 10 15 12" src="https://user-images.githubusercontent.com/1460732/51970764-016b5e80-2478-11e9-9b82-22cfabc3aa3d.png">

Here is our workflow:

1) Author A requires a review from R1, R2, R3
2) Reviewer R1 requires changes, R2 and R3 don't request changes (under Reviewers label, R1 has cross, R2 and R3 have yellow dot)
3) Author A fixes the requested changes and re-request a review from R1 (all reviewers have yellow dot under the Reviewers label)
4) PullReminders still think that author A should do an action
Expected: the action should be back on all the reviewers

The same problem happens, if there is just one reviewer.

This is a private repo, but if you need any more information, just let me know ;)

cc: @plamut

**Comment by abinoda** [comment]
@NejcZupec @plamut @ccpost Thanks for the report. Will look into this today.

**Comment by abinoda** [comment]
@NejcZupec @plamut @ccpost 

I just tested this on a public repo and was not able to reproduce the problem. You can look at the [test PR](https://github.com/nodable/auth-service/pull/18) and/or follow below. I tested this with the setting set to `After one review` as in your examples, then performed the following steps.

1. Opened a new PR and assigned 3 reviewers. The reminder is for reviewers.
2. Requested changes from one of the reviewers (superuser14). The reminder is for the author.
3. Re-requested a review from that reviewer (superuser14). The reminder is for reviewers.

Can you take a look at the test PR and compare it to your own cases to see what might be happening?

**Comment by NejcZupec** [comment]
@abinoda Thank you for very fast response and research ❤️ 

I checked your PR and everything seems correct. I will continue observing the behaviour on our side and let you know, if any discrepancies will pop-up. I know that this was happening in the past.... it looks like the workflow has been already fixed.

Sorry for inconveniences.

**Comment by abinoda** [comment]
@NejcZupec No worries. The logic for this is quite complex so definitely let us know if you observe any issues with it going forward!

**Comment by abinoda** [comment]
@ccpost It turned out there *was* a bug caused when a PR author commented on his or her own PR – we were counting this review toward the total which threw off the logic. We just deployed a fix for this to production so please let us know if you see the issue occur again!

Big thanks also to @NejcZupec @plamut for helping provide information leading to the fix.

**Comment by ccpost** [comment]
@abinoda Apologies I went silent on more debugging info, but thrilled to hear you found the fix.
😁 
This app continues to be one of the best things for daily team workflow we've encountered in a long time!

**Comment by abinoda** [comment]
@ccpost Great to hear! Thanks
